teps/0080-taskrun-prepost-steps.md

```diff
@@ -0,0 +1,163 @@
+---
+status: proposed
+title: 'TaskRun: Pre/Post Steps'
+creation-date: '2021-08-18'
+last-updated: '2021-08-18'
+authors:
+- '@mattmoor'
+---
+
+# TEP-0080: TaskRun: Pre/Post Steps
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes/Caveats (optional)](#notescaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [User Experience (optional)](#user-experience-optional)
+  - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [Implementation Pull request(s)](#implementation-pull-request-s)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+This TEP allows users to partition their list of steps into three distinct phases
+(naming is "to-be-bikeshed"):
+ - `pre-steps`: a section of the TaskRun execution prior to the "real" work.  This
+ would generally consist of "setup" and "download" related work.
+ - `steps`: what we have today.
+ - `post-steps`: a section of the TaskRun execution posterior to the "real" work.
+ This would generally consist of "tear-down" and "upload" related work.
+
+## Motivation
+
+One of the primary motivations for this work is to enable users to help express
+their intent when utilizing "[hermekton](./0025-hermekton.md)" builds.  Hermekton
+lets users disable network access during step execution, but as a `bool` it doesn't
+provide any real leeway to have "pre" steps fetch sources (e.g. `git clone`) or
+dependencies (e.g. `go mod download`) over the network, or "post" steps publish
+results (e.g. `gsutil cp`) outside of that network jail.
```

Comment:
Reading the comment should we have more points here than just `kermekton` ? Especially as "as a `bool` it doesn't provide any real leeway to …" could be seem as something to work on as part of `hermekton` and then as a consequence here.

Reply:
It's certainly most complementary to Hermekton at the moment, but as I mention above (re: TEP-0040) there are ways that this could be quite complementary as a way of expressing `finally`-like semantics (without that necessarily being part of what we do here).

Reply:
The re-usable unit in Tekton is the `Task` - in that sense I think it would be nice to keep `hermekton` features associated to tasks. Cloning a repo and uploading results are things that could be done in tasks specialised for that, and they would not run in hermetic mode.